### PR TITLE
Rip out the atexit handler

### PR DESCRIPTION
**Pull request by DemiMarie**
It causes undefined behavior in multithreaded programs.

### CODE REVIEW

**Comment by Conan-Kudo** [comment]
I'm not sure we can do that, because the point of this handler is to ensure the rpmdb is sane in those scenarios.

**Comment by DemiMarie** [comment]
> I'm not sure we can do that, because the point of this handler is to ensure the rpmdb is sane in those scenarios.

There are two problems with this approach:

1. We need to make sure the rpmdb is reasonable even if RPM dies with SIGKILL.  So the atexit handler cannot be essential for correctness.
2. Other threads may be using RPM while one calls exit().  So we can’t actually free any resources in the atexit() handler.

Now that we use SQLite, I imagine that the rpmdb will *always* be in a consistent state.

**Comment by Conan-Kudo** [comment]
So I guess then, the question would be to @mlschroe if NDB still needs this?

**Comment by cgwalters** [comment]
This seems related to https://github.com/rpm-software-management/rpm/commit/56f49d7f5af7c1c8a3eb478431356195adbfdd25 - in rpm-ostree for example the atomicity of the rpmdb is also done alongside the filesystem too - if we don't have a `SIGINT` handler then we shouldn't have an `atexit` handler either.

**Comment by mlschroe** [comment]
AFACT ndb does not need this. It should be self-healing (and I did write a torture test program to test this).

**Review by Conan-Kudo — APPROVED** [review]

**Comment by pmatilai** [comment]
It should be safe from the db POV to remove this now, yes. However to me this is only a part of a bigger change to remove the associated and now unnecessary signal queue from rpm API and ABI, and there's not a whole lot benefit in doing this part without doing the others (the whole rpm*Rock tracking in rpmdb.c is thread-unsafe) as well. It may also have unexpected side-effects that we'd rather deal with all at once when the signal handler gets purged.

In other words, the change is not wrong as such and will happen sooner or later as a part of a bigger change, but not yet. Thanks for the patch though.
